Fix auto upload

arnohovhannisyan opened the pull request:
#### Short description of what this resolves:
This PR fixes auto-upload by using Chokidar and `vscode.extensions.onDidChange` instead of `vscode.FileSystemWatcher`, which only works on files in the current workspace.

#### Changes proposed in this pull request:

- Use Chokidar instead of `vscode.FileSystemWatcher` to detect changes in user settings
- Use VSC's onDidChange event to check for changes in extensions
- Unlock lock file when returning from the `StartWatch()` function after locking to prevent it from becoming unusable

**Fixes**: 3

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
I have tested it by installing/uninstalling extensions and modifying my settings. I can confirm that the changes don't affect any other areas in the code, and it uploads and downloads successfully.
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution](https://github.com/shanalikhan/code-settings-sync/blob/master/CONTRIBUTING.md#setup-extension-locally) guidelines.
- [ ] My change requires a change to the documentation and GitHub Wiki.
- [ ] I have updated the documentation and Wiki accordingly.

shanalikhan commented:
Thanks for the PR.
I will look deeply in few days and accept after testing.

But if you look into this link.
https://code.visualstudio.com/api/references/vscode-api#workspace.createFileSystemWatcher

It says

> A glob pattern that is applied to the absolute paths of created, changed, and deleted files. Use a relative pattern to limit events to a certain workspace folder.

I think we still can use vscode watcher itself, if they provide watch on absolute paths. If yes, then we only need to fix the paths.

arnohovhannisyan commented:
> A glob pattern that is applied to the absolute paths of created, changed, and deleted files. Use a relative pattern to limit events to a certain workspace folder.

I think what it means by that is you can watch for a specific folder in your

shanalikhan commented:
It was removed on https://github.com/shanalikhan/code-settings-sync/commit/7716e71c316a64519b053a45f5535b648e8a5a53#diff-72050d84a8ab08bacc0bd4ac1c1bfb8f by @nekonenene 

> I think what it means by that is you can watch for a specific folder in your

In 'Workspaces' only or we can set by absolute paths ?

arnohovhannisyan commented:
On that same page, it says this:
> Note that only files within the current workspace folders can be watched.

> Use a relative pattern to limit events to a certain workspace folder.
I'm guessing this means that you can use relative paths to watch a folder, like src or src/utilities, but the rules still apply.

shanalikhan commented:
alright, thanks 
I will merge this into next version.

arnohovhannisyan commented:
This PR has a bug which has been fixed in #814. Please try that one as it not only fixes this bug, but it also adds some great new features.

shanalikhan commented:
Can you re-open this and fix this independent so I can publish the new version containing this only this fix alongwith other PR ?

arnohovhannisyan commented:
Sure. However, It might take a while since I am going to be away from home for a few days. I'll try to fix it as soon as possible.

arnohovhannisyan commented:
I should be able to do it today, but I'm not 100% sure.

shanalikhan commented:
No problem, we can wait. Give me a signal to test and publish when you are ready.

arnohovhannisyan commented:
These changes should fix it. Please look out for formatting issues in the code as I used my old laptop, and it seems like it messed up the formatting a bit. Other than that, it should work.

arnohovhannisyan commented on src/setting.ts:
```diff
@@ -39,6 +35,20 @@ export class CustomSettings {
   ];
   public ignoreUploadFolders: string[] = ["workspaceStorage"];
   public ignoreExtensions: string[] = [];
+  public ignoredItems = [
```
Having a central place for all ignored files and folders lets chokidar handle the ignoring for us. However, I can try to use the existing settings with chokidar's ignored properties.

arnohovhannisyan commented on src/sync.ts:
```diff
@@ -48,36 +52,30 @@ export class Sync {
    * Upload setting to github gist
    */
   public async upload(): Promise<void> {
+    console.log("uploading");
```
Oh yeah, I completely forgot about that. I'll remove as soon as possible.

arnohovhannisyan commented:
Ok, I have reverted to using asynchronous functions, and the code is ready to be further tested.

shanalikhan commented on src/sync.ts:
```diff
@@ -48,16 +50,10 @@ export class Sync {
-        vscode.commands
-          .executeCommand("extension.downloadSettings")
-          .then(() => {
-            if (startUpSetting.autoUpload && tokenAvailable && gistAvailable) {
```
Removing the then part , allows it work as propsosed by adding this line. Have you tested it?

The test case should be it should not hit `autoUpload` when downloading process is completed **through** `autoDownload` and `settings.json` is changed ?

arnohovhannisyan replied:
Yes, I've tested it and that solution is probably better, so I have reverted to using then.

shanalikhan commented on src/setting.ts:
```diff
@@ -30,6 +30,9 @@ export class KeyValue<T, S> {
 
 export class CustomSettings {
   public ignoreUploadFiles: string[] = [
```
can you add `.DS_Store` file also here. Because this file is changed for mac only users and auto upload gets into the loop.
Please include lock file also in this list.

arnohovhannisyan replied:
Sure

shanalikhan commented on src/commons.ts:
```diff
@@ -72,172 +70,26 @@ export default class Commons {
 
-  public async InitiateAutoUpload(path: string): Promise<boolean> {
-    vscode.window.setStatusBarMessage("").dispose();
-    vscode.window.setStatusBarMessage(
```
We need to show user if autoUpload process start processing. Have you moved this status or please add this where the `autoUpload` process is started.

arnohovhannisyan replied:
Should I leave the 5 second delay, or just start uploading right away?

shanalikhan replied:
You can leave 5 second delay.

arnohovhannisyan replied:
Ok, I have done that.

shanalikhan commented on src/service/autoUploadService.ts:
```diff
@@ -0,0 +1,82 @@
+    private options: { en: Environment; commons: Commons; ignored: string[] }
+  ) {
+    vscode.extensions.onDidChange(async () => {
+      if (this.watching) {
```
Thanks for adding. :)
Does new extension adds on `extensions.json` while on upload process?  Previously we needed to restart the code like before.

arnohovhannisyan replied:
After I installed an extension, it auto-uploaded and showed up in my gist.

arnohovhannisyan commented:
Here are the results from my testing
- [x] Run auto upload, while it's processing. Change one observed file and see, does the autoUpload runs again by destroying the old processing state or **waits for it to be complete and ignore the again upload call**.

> Open 2 code windows

I couldn't get two windows open with the development host, but the code still depends on the lock file, so it shouldn't matter how many windows there are open.

shanalikhan commented:
> I couldn't get two windows open with the development host, but the code still depends on the lock file, so it shouldn't matter how many windows there are open.

You can make package from `vsce package` of extension.
Install extension on your computer from compiled and run the code on both windows.

arnohovhannisyan commented:
Ok, I'll try that as soon as I can and update the results when I'm done.

arnohovhannisyan commented:
Ok, I have tested it and can confirm that the lock file works. After starting auto upload on one window, attempting to start it on another window stopped the operation because the lock file was locked.

shanalikhan commented:
Thank you for work.
Its merged and will be released in next version
One more question? Are you available for future bug fixes regarding auto uploads?
I can tag you when where there will be any problem

arnohovhannisyan commented:
Thanks for merging! And yes, I will be available, just tag me and I'll get to work as soon as possible.

shanalikhan commented:
New version is published. :)
Thanks for the PR
